### ISSUE TITLE
Write a proper digital filter for directly connected buttons.

### ISSUE THREAD

**Issue by krisjans** [ShpakovDmitry/My-old-projects#24]
https://github.com/ShpakovDmitry/My-old-projects/blob/6e86ebe9eb74b7c9b303172c605751fa2a4893e7/embedded/EscapeROOM/radio_morse_panel/radio_morse_panel.c#L159-L239

**Comment by ShpakovDmitry** [comment]
I have the following idea regarding digital filter.
Here should be used low-pass-filter with time constant ![изображение](https://user-images.githubusercontent.com/59893335/74671381-b96ac700-51b3-11ea-981e-81f7df970c99.png) much bigger than bounce time ![изображение](https://user-images.githubusercontent.com/59893335/74671422-d4d5d200-51b3-11ea-80bd-e0df4a1e06f4.png) when key is pressed or released, but much lower than "regular" press duration time ![изображение](https://user-images.githubusercontent.com/59893335/74671487-f1720a00-51b3-11ea-871f-bf8ce6b724f8.png).
Simple **IIR** low-pass filter equation would be the following:
![изображение](https://user-images.githubusercontent.com/59893335/74671738-7e1cc800-51b4-11ea-9912-385b0c326ce0.png)
where:
![изображение](https://user-images.githubusercontent.com/59893335/74673109-45322280-51b7-11ea-9fec-e81d7050e73e.png)
![изображение](https://user-images.githubusercontent.com/59893335/74673160-6266f100-51b7-11ea-8cda-64929441f4be.png) - time between button state read,
![изображение](https://user-images.githubusercontent.com/59893335/74673933-ea012f80-51b8-11ea-99b1-f26b3a8ecc5f.png) - current actual state of the button,
![изображение](https://user-images.githubusercontent.com/59893335/74674012-1a48ce00-51b9-11ea-9223-32db9599ab7a.png) - filtered button state,
![изображение](https://user-images.githubusercontent.com/59893335/74674044-30568e80-51b9-11ea-8b61-504f1d5fbe54.png) - previous filtered button state.

Also, using this, it is possible to introduce hysteresis.

I know, this demands floating-point arithmetic, but in this case it is possible to use fixed-point arithmetic.

I would create TIMER OVF ISR which every time ![изображение](https://user-images.githubusercontent.com/59893335/74673160-6266f100-51b7-11ea-8cda-64929441f4be.png) reads button actual state ![изображение](https://user-images.githubusercontent.com/59893335/74673933-ea012f80-51b8-11ea-99b1-f26b3a8ecc5f.png)
calculates and updates button filtered state ![изображение](https://user-images.githubusercontent.com/59893335/74674012-1a48ce00-51b9-11ea-9223-32db9599ab7a.png) and according to hysteresis send "signal" to method which takes an action on button state change.

#### Code example:
```c
ISR(TIMER_OVF){
  // read actual button state through PINB
  // set the flag
  STATE_UPDATED_flag = true;
}

void UpdateButtonState(void){
  if(STATE_UPDATED_flag){
    STAE_UPDATED_flag = false;
    // calculate new filtered button states
    // update button state using hysteresis and filtered states
    // if any button state changed put it in FIFO or
    // better in cyclic buffer for further button sequence analysis
  }
}

void CheckValidPressSequence(void){
  // ...
  if(correct)
    PlayMorsePhrase();
  else
    SwitchOnRedLed();
}

```

This will change the code a lot, but i think that according to Issue #14, code anyway should be changed a lot.

Also this IIR filter implementation maybe more complicated than waiting while time elapsed from last de-bounce event is larger than de-bounce duration.

**Comment by krisjans** [comment]
You might want to fix issue #14 before this one. First make ISR fast and non-blocking and after that introduce proper filter.

**Comment by ShpakovDmitry** [comment]
@krisjans 

Please, can you give an advice on the following.
Maybe it is better to create new branch called `dev`, with infinite lifetime, and in future merge branches that resolves issues into `dev` branch? And when everything works fine merge it into `master` branch and continue work on `dev` branch.
This is due to the following. I think there might be the situation when issue is successfully solved, code may become, lets say, in `transition` state, when it is not fully working and demands new feature to be added. So, we develop code in `dev` branch until it starts to work and then `release` it by merging into `master` branch. Thus it is possible to merge into `master` branch only in situations when code is fully functional. And thus, it is possible to access to functional code(`master`) at any time.
Correct me, please, if i am wrong regarding development flow.

**Comment by krisjans** [comment]
@ShpakovDmitry
Yes, just like coding guidelines - it is about taste and convenience - use whatever you like.

But - please take a look how other teams deals with this problem. Google for "git branch strategy" - you will find interesting articles, diagrams and videos about this topic.

In my opinion, general "rule" is that "master" is the main branch where development happens. Developers either commit directly to the master OR create feature branch for every bug or feature  and when it is done merge it back to master. Inside feature branch you may solve several issues and even have non-working or unstable states.
Release is either "tag" in master or "release branch" from master.

P.S. I see that you create branch names  to match issue numbers. But this makes things pretty confusing and messy. Maybe give a branch meaningful name (like "add feature X") and later link it with issue(s). And when you merge branch back to master, you might want to consider "rebase" strategy. 
After all - git was invented to help you understand what you did 6 months (or 6 years) ago and why.

**Comment by ShpakovDmitry** [comment]
Ok, actually, i had not thought, that for non-working and unstable states it is possible to use the same feature branch. Will use this kind of flow at the moment.

P.S. Yeah, i have noticed, recently, that branch names with issue number only, are confusing and demands extra time to check what does this branch purpose really mean, which is unnecessary action. Now, i can imagine how it may be confusing if work is done on several issues at one time.
So for this purpose in last pull request a meaningful branch name was used ` coding_style_md `
